crates/factor-outbound-networking/src/runtime_config.rs

```diff
@@ -0,0 +1,250 @@
+#[cfg(feature = "spin-cli")]
+pub mod spin;
+
+use std::{collections::HashMap, str::FromStr, sync::Arc};
+
+use anyhow::{ensure, Context};
+use rustls::{ClientConfig, RootCertStore};
+use rustls_pki_types::{CertificateDer, PrivateKeyDer};
+
+/// Runtime configuration for outbound networking.
+#[derive(Debug)]
+pub struct RuntimeConfig {
+    // Maps component ID -> HostClientConfigs
+    component_host_client_configs: HashMap<String, HostClientConfigs>,
+    default_client_config: Arc<ClientConfig>,
+}
+
+// Maps host authority -> ClientConfig
+type HostClientConfigs = Arc<HashMap<String, Arc<ClientConfig>>>;
+
+impl RuntimeConfig {
+    pub fn new(tls_configs: impl IntoIterator<Item = TlsConfig>) -> anyhow::Result<Self> {
+        let mut component_host_client_configs = HashMap::<String, HostClientConfigs>::new();
+        for tls_config in tls_configs {
+            ensure!(
+                !tls_config.components.is_empty(),
+                "client TLS 'components' list may not be empty"
+            );
+            ensure!(
+                !tls_config.hosts.is_empty(),
+                "client TLS 'hosts' list may not be empty"
+            );
+            let client_config = Arc::new(tls_config.to_client_config()?);
+            for component in &tls_config.components {
+                let host_configs = component_host_client_configs
+                    .entry(component.clone())
+                    .or_default();
+                for host in &tls_config.hosts {
+                    validate_host(host)?;
+                    // First matching (component, host) pair wins
+                    Arc::get_mut(host_configs)
+                        .unwrap()
+                        .entry(host.clone())
+                        .or_insert_with(|| client_config.clone());
+                }
+            }
+        }
+
+        let default_client_config = Arc::new(TlsConfig::default().to_client_config()?);
```

Comment:
Nit: I'm not sure that a `Default` impl for `TlsConfig makes a ton of sense considering we treat empty lists of components or hosts as errors elsewhere. 

Instead I think it makes more sense to just have a function for getting a default client config. I recognize this would require refactoring `to_client_config` and you might not want to bother, but I do think it makes the code easier to understand.

Reply:
It's just to enable the `TlsConfig { ..., ..Default::default() }` idiom :shrug: